swe_roe/ex2.c

```diff
@@ -8,418 +8,461 @@ static char help[] = "Partial 2D dam break problem.\n";
-  PetscInt *cell_offset; /* offset for internal cell IDs that share the vertex              */
-
-  PetscInt *edge_ids; /* edge IDs that share the vertex                       */
-  PetscInt *cell_ids; /* internal cell IDs that share the vertex              */
-
-  RDyCoordinate *coordinate; /* (x,y,z) location of the vertex                       */
-} RDyVertex;
-
-typedef struct {
-  PetscInt *id;        /* id of the edge in local numbering         */
-  PetscInt *global_id; /* global id of the edge in local numbering */
-
-  PetscBool *is_local; /* true if the edge : (1) */
-                       /* 1. Is shared by locally owned cells, or   */
-                       /* 2. Is shared by local cell and non-local  */
-                       /*    cell such that global ID of local cell */
-                       /*    is smaller than the global ID of       */
-                       /*    non-local cell */
-
-  PetscInt *num_cells;  /* number of faces that form the edge        */
-  PetscInt *vertex_ids; /* ids of vertices that form the edge        */
-
-  PetscInt *cell_offset; /* offset for ids of cell that share the edge */
-  PetscInt *cell_ids;    /* ids of cells that share the edge          */
-
-  PetscBool *is_internal; /* false if the edge is on the mesh boundary */
-
-  RDyVector     *normal;   /* unit normal vector                        */
-  RDyCoordinate *centroid; /* edge centroid                             */
+  /// numbers of cell vertices
+  PetscInt *num_vertices;
+  /// numbers of cell edges
+  PetscInt *num_edges;
+  /// numbers of cell neigbors (themselves cells)
+  PetscInt *num_neighbors;
+
+  /// offsets of first cell vertices in vertex_ids
+  PetscInt *vertex_offsets;
+  /// IDs of vertices for cells
+  PetscInt *vertex_ids;
+
+  /// offsets of first cell edges in edge_ids
+  PetscInt *edge_offsets;
+  /// IDs of edges for cells
+  PetscInt *edge_ids;
+
+  /// offsets of first neighbors in neighbor_ids for cells
+  PetscInt *neighbor_offsets;
+  /// IDs of neighbors for cells
+  PetscInt *neighbor_ids;
+
+  /// cell centroids
+  RDyPoint *centroids;
+  /// cell areas
+  PetscReal *areas;
+} RDyCells;
+
+/// Allocates and initializes an RDyCells struct.
+/// @param [in] num_cells Number of cells
+/// @param [out] cells A pointer to an RDyCells that stores allocated data.
+///
+/// @return 0 on success, or a non-zero error code on failure
+PetscErrorCode RDyCellsCreate(PetscInt num_cells, RDyCells *cells) {
+  PetscFunctionBegin;
 
-  PetscReal *length; /* length of the edge                        */
+  PetscInt vertices_per_cell  = 4;
+  PetscInt edges_per_cell     = 4;
+  PetscInt neighbors_per_cell = 4;
+
+  PetscCall(RDyAlloc(PetscInt, num_cells, &cells->ids));
+  PetscCall(RDyAlloc(PetscInt, num_cells, &cells->global_ids));
+  PetscCall(RDyAlloc(PetscInt, num_cells, &cells->natural_ids));
+  RDyFill(PetscInt, cells->global_ids, num_cells, -1);
```

Comment:
It looks a little odd that there isn't a return value from `RDyFill`. How about the following?

```

#define RDyFill(type, array, count, value) \
  for (size_t i = 0; i < (count); ++i) {   \
    *(array+i) = (type)value;                \
  }


RDyFillInt(PetscInt *array, PetscInt count, PetscInt value) {
  PetscFunctionBegin;
  RDyFill(PetscInt, array, num_cells, value);
  PetscErrorCode(0);
}

PetscCall(RDyFillInt(cells->global_ids, num_cells, -1));
```

Reply:
That's fine with me, if we'd like everything to have a return code. I was motivated by `PetscOptionsBegin` and `PetscOptionsEnd`, which are also macros that don't return error codes, so I don't think it's a big deal either way stylistically.

Reply:
In that case, let's have a return code. Everything else looks good.

Reply:
I hacked a return code into the macro. This operation always succeeds unless you give it bad memory, which is very easy to troubleshoot.